File: CS/ExpressionEvaluator Filter/FilteredDataSupport.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.ComponentModel;
using System.Windows;
using System.Collections;
using DevExpress.Xpf.Grid;
using DevExpress.Data.Filtering.Helpers;

namespace ExpressionEvaluator_Filter {
    public class FilteredDataSupport {
        public static readonly DependencyProperty VisibleDataProperty =
            DependencyProperty.RegisterAttached("VisibleData", typeof(IList), typeof(FilteredDataSupport), new PropertyMetadata(OnVisibleDataChanged));

        public static void SetVisibleData(UIElement element, IList value) {
            element.SetValue(VisibleDataProperty, value);
        }
        public static IList GetVisibleData(UIElement element) {
            return (IList)element.GetValue(VisibleDataProperty);
        }

        private static void OnVisibleDataChanged(DependencyObject sender, DependencyPropertyChangedEventArgs e) {
            GridControl grid = sender as GridControl;
            if(grid == null)
                return;
            if(e.OldValue == null && e.NewValue != null)
                grid.FilterChanged += OnFilterChanged;
            else if(e.OldValue != null && e.NewValue == null)
                grid.FilterChanged -= OnFilterChanged;
        }

        static void OnFilterChanged(object sender, RoutedEventArgs e) {
            GridControl grid = sender as GridControl;
            if(grid == null)
                return;
            var res = grid.DataController.GetAllFilteredAndSortedRows();
            IList visibleData = grid.GetValue(VisibleDataProperty) as IList;
            if(visibleData == null)
                return;
            visibleData.Clear();
            foreach(object item in res) {
                visibleData.Add(item);
            }
        }
    }
}

File: VB/ExpressionEvaluator Filter/FilteredDataSupport.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.ComponentModel
Imports System.Windows
Imports System.Collections
Imports DevExpress.Xpf.Grid
Imports DevExpress.Data.Filtering.Helpers

Namespace ExpressionEvaluator_Filter
	Public Class FilteredDataSupport
        Public Shared ReadOnly VisibleDataProperty As DependencyProperty = DependencyProperty.RegisterAttached("VisibleData", GetType(IList), GetType(FilteredDataSupport), New PropertyMetadata(AddressOf OnVisibleDataChanged))

		Public Shared Sub SetVisibleData(ByVal element As UIElement, ByVal value As IList)
			element.SetValue(VisibleDataProperty, value)
		End Sub
		Public Shared Function GetVisibleData(ByVal element As UIElement) As IList
			Return CType(element.GetValue(VisibleDataProperty), IList)
		End Function

		Private Shared Sub OnVisibleDataChanged(ByVal sender As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			Dim grid As GridControl = TryCast(sender, GridControl)
			If grid Is Nothing Then
				Return
			End If
			If e.OldValue Is Nothing AndAlso e.NewValue IsNot Nothing Then
				AddHandler grid.FilterChanged, AddressOf OnFilterChanged
			ElseIf e.OldValue IsNot Nothing AndAlso e.NewValue Is Nothing Then
				RemoveHandler grid.FilterChanged, AddressOf OnFilterChanged
			End If
		End Sub

		Private Shared Sub OnFilterChanged(ByVal sender As Object, ByVal e As RoutedEventArgs)
			Dim grid As GridControl = TryCast(sender, GridControl)
			If grid Is Nothing Then
				Return
			End If
			Dim res = grid.DataController.GetAllFilteredAndSortedRows()
			Dim visibleData As IList = TryCast(grid.GetValue(VisibleDataProperty), IList)
			If visibleData Is Nothing Then
				Return
			End If
			visibleData.Clear()
			For Each item As Object In res
				visibleData.Add(item)
			Next item
		End Sub
	End Class
End Namespace
